docs/src/lens.md

```diff
@@ -1,149 +1,251 @@
 
-```rust,noplaypen
-{{#include ../book_examples/src/lens_md.rs:todo_item}}
+## Fundamentals: Definition and Implementation
+
+Like Rust itself, lenses are one of those things that require effort up front to learn, but are very fun and effective to use once you understand them. This section represents the effort part of the equation. I promise if you stick with it you will reap the rewards.
+
+### Definition
+
+Let's start with the definition of a `Lens`:
+
+```rust
+pub trait Lens<T, U> {
+    fn with<F: FnOnce(&U)>(&self, data: &T, f: F);
+
+    fn with_mut<F: FnOnce(&mut U)>(&self, data: &mut T, f: F);
+}
 ```
 
-We would like our widget to display the title of the item, and then below
-that to display two checkmarks that toggle the 'completed' and 'urgent' bools.
-`Checkbox` (a widget included in Druid) implements `Widget<bool>`.
-How do we use it with `TodoItem`? By using a `Lens`.
+I've copied this definition from the `druid` source code, but then simplified it a little, by removing the return types, as they are not fundamental to the way lenses work.
 
-## Conceptual
+The first thing to notice is the generics on the `Lens` itself. There are 3 types involve in the lens: the lens itself, `T` and `U`. The two type parameters represent the mis-match that lenses solve: we have a function that operates on `U`, and an object of type `T`, so we need to transform `T` into `U` somehow.
 
-You can think of a lens as a way of "focusing in" on one part of the data. You
-have a `TodoItem`, but you *want* a `bool`.
+### Implementation
 
-`Lens` is a trait for types that perform this "focusing in" (aka *lensing*).
-A simplified version of the `Lens` trait might look like this:
+Time for an example. Let's implement & use `Lens` manually so we can see what's going on.
 
-```rust,noplaypen
-{{#include ../book_examples/src/lens_md.rs:simple_lens}}
+```rust
+struct Container {
+    inner: String,
+    another: String,
+}
+
+// Here the lens doesn't have any data, but there are cases where
+// it might, for example it might contain an index into a collection.
+struct InnerLens;
+
+// Our lens will apply functions that operate on a `String` to a `Container`.
+impl Lens<Container, String> for InnerLens {
+    fn with<F: FnOnce(&String)>(&self, data: &Container, f: F) {
+        f(&data.inner);
+    }
+
+    fn with_mut<F: FnOnce(&mut String)>(&self, data: &mut Container, f: F) {
+        f(&mut data.inner);
+    }
+}
 ```
 
-That is, this type takes an instance of `In`, and returns an instance of `Out`.
+This is a very simple case. All we need to do is project the function onto the field. Notice that this isn't the only vaid lens from `Container` to `String` we could have made - we could also project from `Container` to `another`. We made the choice how to transform `Container` into `String` when we implemented `Lens`.
```

Comment:
```suggestion
This is a very simple case. All we need to do is project the function onto the field. Notice that this isn't the only valid lens from `Container` to `String` we could have made - we could also project from `Container` to `another`. We made the choice how to transform `Container` into `String` when we implemented `Lens`.
```

I think I would just use a single field with the first example? I like the idea of keeping it maximally simple.

Reply:
I don't quite understand what "project the function onto the field" means here. I guess you're using project as a synonym for `map`? Like map a field to another field? I only just thought of this, but Lensing is like mapping from one thing to another right?

Reply:
I agree that maybe 'project' is a bit too mathy? @arthmis you're correct, you can think of a lens as a two-way map; you can get one value from another, and then mutations in that value are also reflected in the original value.